Chore: Update jsdoc plugin and tweak rules in effect

brettz9 opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

- Update `eslint-plugin-jsdoc` peer dependenciy in `eslint-config-eslint` (and update as dev. dep. in `eslint`)
- Update `.eslintrc.js` to disable rules that files with `@example` would otherwise be wrongly flagged
- Indicate rules applied from `plugin:jsdoc/recommended` (but switched to `error`'s), add other helpful jsdoc rules 
- Fix JSDoc per latest JSDoc plugin update

#### Is there anything you'd like reviewers to focus on?

- One minor JSDoc change from semi-colon to comma should not be necessary, but waiting on a fix to `jsdoc-type-pratt-parser` and disabling linter around a few TypeScript constructs which the parser is also awaiting support for
- The `function() : string` syntax is not TypeScript-based, so switched to that indicated by TS
- For the sake of a rule to avoid accidental extra asterisks, switched some initial line asterisks to `-`

Note that I commented out the following [jsdoc](https://github.com/gajus/eslint-plugin-jsdoc) rules as they would have entailed more diff noise, and I wasn't sure if you wanted them, or how, but I do think they'd be good to enable:

- `jsdoc/check-indentation` - This and the next rule might add noise, and not sure you were consistent, but might be nice to enforce consistency 
- `jsdoc/check-line-alignment`
- `jsdoc/require-file-overview` - You have these in use already, but would need to confine to a directory
- `jsdoc/require-throws` - Should be useful to require denoting the types thrown by the function when a `throw` is found
- `jsdoc/check-values` - This rule can enforce SPDX license identifiers; if ok, I could add the other text currently there to a `@copyright` tag

bmish commented on lib/linter/report-translator.js:
```diff
@@ -32,18 +32,18 @@ const interpolate = require("./interpolate");
+ * @property {number} column The column number
+ * @property {(number|undefined)} [endLine] The ending line number
+ * @property {(number|undefined)} [endColumn] The ending column number
+ * @property {(string|null)} nodeType Typeo f node
```
typo

brettz9 replied:
Thanks--you caught a literal "typeo" :-) !  Fixed...

bmish commented on lib/rules/no-loop-func.js:
```diff
@@ -174,7 +174,7 @@ module.exports = {
          * - has any references which refers to an unsafe variable.
          * @param {ASTNode} node The AST node to check.
-         * @returns {boolean} Whether or not the node is within a loop.
+         * @returns {void}
```
This is incorrect. Should be boolean but we should update `return` on line 183 to `return false` to make it always return a boolean.

brettz9 replied:
Updating it on line 183 to return false still does not cause a boolean to be returned everywhere, as the function doesn't return elsewhere--i.e., it effectively returns `undefined` elsewhere too.

If this were a generic utility, it might make sense to return a boolean, but the `checkForLoops` function is currently only used as an ESLint rule callback and is not used for validation, so despite the name of the function, there is no benefit for it to return a boolean.

bmish replied:
Ah my bad, I misread the code, you're right.

brettz9 commented:
I see this PR might need to wait on `eslint` and `eslint-config-eslint` updating `engines` to Node 12+...

nzakas commented:
The engine upgrade won’t happen until the next major version (which will hopefully be soon).

mdjermanovic commented on .eslintrc.js:
```diff
@@ -76,6 +76,22 @@ module.exports = {
         "internal-rules/multiline-comment-style": "error"
     },
     overrides: [
+        {
+            files: "**/no-useless-escape.md/*.js",
+            rules: {
+                "no-unused-expressions": "off"
+            }
+        },
+        {
+            files: "**/ast-utils.md/*.js",
+            rules: {
+                "no-floating-decimal": "off",
+                "no-octal": "off",
+                "no-unused-expressions": "off",
+                "no-multi-spaces": "off",
+                semi: "off"
+            }
+        },
```
It looks like linting jsdoc examples will introduce more problems than benefits if we'll have to override configuration often.

brettz9 replied:
While some of the rules could be loosened across all examples, especially in ESLint where you often actually want some bad practices showing up in examples (e.g., the `no-octal` rule there)--as well as the bad practices shown in Markdown example blocks, if you wanted to lint those files' examples, yes, it would involve more frequent tweaking like this.

In my own projects, I have tended to consider the cost worthwhile in ensuring that the examples not only dogfood styles, but work as expected, but I admit it comes at something of a cost (esp. since we have no mechanism similar to the Markdown plugin whereby one can add disable directives just above the jsdoc blocks instead of within them (which would be ugly) or within an external file (which gets less maintainable being out-of-band)).

I can disable the `check-examples` rule then if desired.

nzakas replied:
Yeah, I don’t think we are going to get a ton of benefit from linting examples, so let’s keep it off for now. We can always revisit later.

brettz9 commented:
Made a number of mostly small, misc. changes.

A few items/questions.

1. I added the helpful `eslint-comments` plugin, but it's perhaps most useful rule, `eslint-comments/require-description` is currently commented out as it would require 136 changes in this repo alone.

This rule requires that a reason be given for disabling the linting--something very helpful both to understand why there was some laxening, and to prevent wasted time of other devs on seeing if the disable directive could be removed.

Let me know if you wish to enable this.

2. For the rule `jsdoc/check-line-alignment`, enforcing the "never" vs. "always" style would require 126 vs. 1833 changes. I'd go with "never", but wanted to confirm. In other words, this would not require or even allow this type of extra spacing between docs:

```js
/**
 * My function.
 *
 * @param {string} lorem  Description.
 * @param {int}      sit        Description multi words.
 */
```

but would instead insist on:

```js
/**
 * My function.
 *
 * @param {string} lorem Description.
 * @param {int} sit Description multi words.
 */
```

3. There were 37 files without a `@fileoverview`. I could enable `jsdoc/require-file-overview: "error"` or enable on a particular directory if you wanted to ensure each file or each file in a particular directory had this. If you don't care if you forget it yourself or if others do, I can ignore, but I mention as this was a frequent pattern and mentioned in the docs.

4. `jsdoc/require-throws: "error"` 

I went ahead and enabled this one, fixing 42 errors. Let me know if you want me to revert that commit. It can, of course, only detect when there is a `throw` statement.

5. I added some extra enforcement of jsdoc types, e.g., requiring that any documented `Promise` have a type, that the more TypeScript-like "any" be used over "*" and that the older dotted `Object.<something>` notation be changed to `Object<something>`.

One additional change I prefer in my own projects (and currently commented out in the ESLint config) is to prohibit use of a plain `Array`, `Object`, or `Function`, except as a `typedef` base type. One of the most disappointing experiences in perusing code docs in my experience is finding such a type as it often gives very little idea of what detailed structure the type has and is usually the type one most needs to understand.

In rarer cases, such as with a very general purpose utility, there may truly be no limits on the specific type of array, object, or function that may be returned, but in such cases, one can insist on a few types you might create `@typedef`s for, such as `GenericArray`, `GenericObject`, and `GenericCallback`.

Making this change will require a large amount of work, however (36, 398, and 54 fixes, respectively, just in this repo). I could disable this rule for now on the `eslint` project, but encourage it on other projects by adding it by default to `eslint-config-eslint`.

brettz9 commented:
I've also added a commit for Unicorn, also with a good number of changes.

nzakas commented:
Our challenge here is that we have a large amount of outside contributors, and we don’t want to put in too many rules that enforce things that aren’t auto fixable. That’s too discouraging for new or infrequent contributors. 

So checking alignment will be particularly annoying if it can’t be autofixed. If it can be autofixed, then let’s do it, but if not, I’d prefer not enforcing any type of alignment.

The others seem like good ideas and I’d suggest doing those each as a separate PR so we can deal with any issues that come up in a more focused way.

I’d also prefer we allow Object, Array, etc. I don’t really want a lot of energy spent on figuring out the correct type. Getting people to include JSDoc at all was really hard, I don’t want to throw up any more barriers than necessary

brettz9 commented:
This has been rebased, and I think is ready again for review now that `master` has updated `engines`. Note, however, that I removed the Unicorn addition here, in part because it is currently failing, and in part because I think the PR was already getting crowded, so I could see about submitting that in a separate PR.

bmish commented on packages/eslint-config-eslint/default.yml:
```diff
@@ -36,30 +72,67 @@ rules:
     eol-last: "error"
     eqeqeq: "error"
+
+    eslint-comments/disable-enable-pair: ["error", {allowWholeFile: true}]
```
Just FYI, it's generally better practice to use `/* eslint foo:"off" */` to disable a rule for an entire file than to use the disable/enable comments around the entire file.  

I am adding an explanation about this here: https://github.com/mysticatea/eslint-plugin-eslint-comments/pull/60

nzakas replied:
Agree

brettz9 replied:
Sounds good. Added a commit.

bmish commented on packages/eslint-config-eslint/default.yml:
```diff
@@ -125,39 +198,41 @@ rules:
     node/no-new-require: "error"
     node/no-path-concat: "error"
-    node/no-process-exit: "error"
+    node/no-process-exit: "error" # recommended
```
Why are recommended rules enabled here again when we already enable the recommended configs that include them?

brettz9 replied:
In the file, there are two cases where recommended rules are repeated:

1. When overriding a behavior of "warn" to change to "error"
2. Such as the example you cite where I just added the comment for those lines which can indeed just be removed if desired (I didn't know though if someone had it in mind to hard-code the rules in case the recommendations for those particular rules changed, but would be fine to remove them too.)

brettz9 replied:
I've gone ahead and added a commit to remove the rules which are already included by the recommended configs.
